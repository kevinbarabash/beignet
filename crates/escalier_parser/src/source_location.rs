use std::cmp::{max, min};
use std::ops::Range;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Position {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

pub type Span = Range<usize>;

pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

pub fn merge_spans(left: &Span, right: &Span) -> Span {
    Span {
        start: min(left.start, right.start),
        end: max(left.end, right.end),
    }
}
