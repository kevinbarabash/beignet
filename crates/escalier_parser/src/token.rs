use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
