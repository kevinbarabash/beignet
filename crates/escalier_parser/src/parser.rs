use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
