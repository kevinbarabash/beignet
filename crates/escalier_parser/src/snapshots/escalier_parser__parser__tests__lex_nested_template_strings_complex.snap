---
source: crates/escalier_parser/src/parser.rs
expression: "tokens[0].kind"
---
StrTemplateLit {
    parts: [
        Token {
            kind: StrLit(
                "ids = ",
            ),
            span: 0..7,
        },
        Token {
            kind: StrLit(
                "",
            ),
            span: 49..50,
        },
    ],
    exprs: [
        Expr {
            kind: Call {
                args: [
                    Expr {
                        kind: Literal(
                            String(
                                ", ",
                            ),
                        ),
                        span: 43..47,
                    },
                ],
                callee: Expr {
                    kind: Member {
                        object: Expr {
                            kind: Call {
                                args: [
                                    Expr {
                                        kind: Function {
                                            params: [
                                                FuncParam {
                                                    pattern: Pattern {
                                                        span: 21..23,
                                                        kind: Ident(
                                                            BindingIdent {
                                                                name: "id",
                                                                span: 21..23,
                                                                mutable: false,
                                                            },
                                                        ),
                                                    },
                                                    type_ann: None,
                                                    optional: false,
                                                },
                                            ],
                                            body: Expr(
                                                Expr {
                                                    kind: TemplateLiteral {
                                                        parts: [
                                                            String(
                                                                "x",
                                                            ),
                                                            String(
                                                                "",
                                                            ),
                                                        ],
                                                        exprs: [
                                                            Expr {
                                                                kind: Identifier(
                                                                    "id",
                                                                ),
                                                                span: 32..34,
                                                            },
                                                        ],
                                                    },
                                                    span: 27..36,
                                                },
                                            ),
                                            type_ann: None,
                                        },
                                        span: 17..36,
                                    },
                                ],
                                callee: Expr {
                                    kind: Member {
                                        object: Expr {
                                            kind: Identifier(
                                                "ids",
                                            ),
                                            span: 9..12,
                                        },
                                        property: Expr {
                                            kind: Identifier(
                                                "map",
                                            ),
                                            span: 13..16,
                                        },
                                    },
                                    span: 9..16,
                                },
                            },
                            span: 9..37,
                        },
                        property: Expr {
                            kind: Identifier(
                                "join",
                            ),
                            span: 38..42,
                        },
                    },
                    span: 9..42,
                },
            },
            span: 9..48,
        },
    ],
}
